### PR TITLE
Ui modes (kiosk, kid, full)

### PR DESCRIPTION
**Pull request by zigurana**
Hi, after a beta-testing period of a couple of weeks, where no bugs have been reported, I would like to offer these enhancements to the RetroPie/EmulationStation branch. See also[ this thread](http://blog.petrockblock.com/forums/topic/kid-friendly-retropiees-ui-modes-favorites-hiding-items-s-testers-wanted/) on the forum.

Added features:
- UI mode can be set to Full (status quo), Kiosk (show only items that are not hidden explicitly), Kid (show only items that are explicitly whitelisted).[ link](http://blog.petrockblock.com/forums/topic/kid-friendly-retropiees-ui-modes-favorites-hiding-items-s-testers-wanted/#post-111729)
- Configuration items are UI mode depended, so ppl can set the system to Kiosk or Kid without having to worry about config changes being tampered with.
- Theming now allows the addition of Hidden / Kidmode and Favorites as icons in the theme. This also enables the quick toggling of these booleans using x/y for favorite and kidgame resp. [link to post](http://blog.petrockblock.com/forums/topic/kid-friendly-retropiees-ui-modes-favorites-hiding-items-s-testers-wanted/page/2/#post-114024)
- Random game selection (aka: surprise me!) Given the above filters/constraints, select a random game. [Forum post](http://blog.petrockblock.com/forums/topic/kid-friendly-retropiees-ui-modes-favorites-hiding-items-s-testers-wanted/page/2/#post-115616)

Combined with the latest improvements for the Runcommand in RetroPie 3.5, I feel these changes will allow users to create a well protected system (for use with kids/guests), while still having the fun of configuration tuning :-).

Please consider this PR, and let me know if you need anything.
Also, if you need me to 'squash' these commits before creating a PR, let me know. I will figure out how to do it, and resubmit.

### CODE REVIEW

**Comment by zigurana** [comment]
Bug found in this PR: http://blog.petrockblock.com/forums/topic/kid-friendly-retropiees-ui-modes-favorites-hiding-items-s-testers-wanted/page/2/#post-116839

**Comment by joolswills** [comment]
Thanks - I'm not convinced on including such large changes into our ES - the more that gets added the more that needs maintaining, and I think it's best to handle this in your own fork (I would consider adding it as an optional module in retropie - there is a PR for that but that needs de-duplicating - I can look at that when I have some time)

I would also note that the changes above include looks like they need cleaning up and there are whitespace errors in the changes. and commits then reverts which I think should be cleaned up a bit - perhaps squashing some of the commits etc also. Most importantly there are currently conflicts anyway so it can't be merged as is.

**Comment by HerbFargus** [comment]
I know that we don't want to maintain too much and don't want to make too many major changes from upstream in hopes upstream will continue development, but at least for the favourites function it may be worth including in our fork  (as it has been requested a lot from users) which would eliminate some code duplication with this pull request as well.  

At the very least I would agree that having it as a secondary module would be good for those that want those options.

**Comment by zigurana** [comment]
Hey, thanks for considering this, and its a good thing we get a
conversation going.

I agree that the PR needs to be cleaned up before it could ever be
included. Part of the conflicts have been introduced with release 3.5
(after i had started working on background music) and I will incorporate
those on my side. When I submitted the PR, there were no conflicts,  I
swear!
In any case, this is my first PR ever, so I'll need to figure out how to
properly serve out these changes, by squashing the commits etc. Whitespace
can be cleaned up easily enough.

How large do you guys estimate the chance of Aloshi picking up the reigns
again? There has been very little activity from his side for at least half
a year now. The maintanance you mention only  occurs when upstream
development actually happens.

What I am trying to find out is this: to what extend is ES in RetroPie
going to be developed further? Are you only going to allow bugfixes, or can
we expand ES to allow for instance better theming?
I have some other plans lined up and would appreciate some idea of the
chance of eventual uptake into the RetroPie fork.

Are there plans to switch another frontend, one that is still under
development?

Thanks again,
Daniel

Op vr 12 feb. 2016 08:57 schreef HerbFargus <notifications@github.com>:

**Comment by petrockblog** [comment]
I also would suggest that we add some ES features via well-tested (!) pull-requests.

However, I also have the feeling that it might be a good move to move to another front-end in the long term. In the best case that front-end uses the same configuration interface that ES is using and offering now. EmulationStation is very hard to maintain and test. Maybe a script language like Python or Lua would be a better choice for a new, easier to maintain and extend front-end.

**Comment by joolswills** [comment]
I am planning to look at attractmode - we could then use inputstation to do the controller config etc.

**Comment by joolswills** [comment]
We could of course offer up both frontends/ I am still hopeful that Aloshi will continue development at some point - it's still been less than a year, and I have seen much longer gaps between development of projects in the past (including my own).

**Comment by zigurana** [comment]
Ok, think the arguments are clear. I will update and clean up this PR. Then I will resubmit it here and to the upstream Aloshi branch. At that point you can decide what to do with it. 
Alternatively, I would be fine with maintaining my own fork (updating it with the latest ES commits), and have people who are interested installing it via the experimental scripts. That way it would be zero additional effort for you guys, while still reaching many.

**Comment by joolswills** [comment]
sounds good - posting it upstream would be useful.

I think I would be happy with the experimental script and you maintaining it for now, but having a clean patchset would be useful - we can see if anything changes in ES development and decide what to do later.

**Comment by zigurana** [comment]
Cool. One question before I close this PR: Jools said:
"I would consider adding it as an optional module in retropie - there is a
PR for that but that needs de-duplicating"

How would that approach be different from remaining in the experimental
menu? And what do you mean exactly by de-duplicating? Is that something I
need to prepare in my fork to  facilitate?

Thanks,
Daniel



Op zo 14 feb. 2016 21:26 schreef Jools Wills <notifications@github.com>:

**Comment by joolswills** [comment]
it would still be in experimental, just the script module for retropie-setup to install it that gizmo put up duplicates code from the main emulationstation module - so I'm referring to that. I will look at it when I have some time.

**Comment by joolswills** [comment]
Which branch should I be using to build from your fork - master ?

I have a module ready to commit, but it relies on your branch including our changes - or at least this one https://github.com/RetroPie/EmulationStation/pull/17 - so that the launching is functionally the same - you could probably merge in all our changes - but you may need to resolve a conflict or two.

**Comment by zigurana** [comment]
I'll need to have a look at it when I get home tonight. (i am at GMT +1, so
in about 7 hours)
Lets say I can get it done tonight, would that be ok?

You can take the fork called "UI_MODES..."
Thanks,
Daniel


Op ma 15 feb. 2016 10:07 schreef Jools Wills <notifications@github.com>:

**Comment by joolswills** [comment]
No rush - whenever you want/have time :)

**Comment by zigurana** [comment]
Hey, I've updated my branch to be on par with the RetroPie/EmulationStation/master branch.
Build and runs well.

Nice implementation for the file deletion!

You mentioned white-space errors, could you point out some, so know what you mean exactly.
Moreover, what spacing would you like to see? Where I work, anything else than "1 tab get converted into 4 spaces" is strictly verboten. However, I noticed that the github website shows something like tab=8 spaces.

Next commit I will try to do via a rebase action, so it should be squashed into a single commit (or a couple, splitting out between my work and that of Kaptainka before me).

**Comment by joolswills** [comment]
well, you should stick to whatever the indentation in ES is in this case - but I meant there were some additional tabs on blank lines etc and some of the indentation looked odd compared to the other code afair

File: es-app/src/FileData.cpp [hunk 1]
```diff
@@ -142,3 +176,21 @@ void FileData::sort(const SortType& type)
...
 }
+
+FileData* FileData::getRandom(bool filterHidden, bool filterFav, bool filterKid) const
+{
+	LOG(LogDebug) << "FileData::getRandom("<< filterHidden << ", " << filterFav << ", " << filterKid << ")";
+	
+	//Get list of files
+	std::vector<FileData*> list = getFilesRecursive(GAME,filterHidden, filterFav, filterKid);
+	const unsigned long n = list.size();
+    LOG(LogDebug) << "   found games: " << n;
```
Comment (joolswills) [inline es-app/src/FileData.cpp]: this line for example.

**Comment by joolswills** [comment]
I went to sort the module out but you seem to have two branches now 

UI_modes_Kiosk_Kid_Full and
origin/UI_modes_Kiosk_Kid_Full

the first one is where this PR came from, but the other one is the one you have been updating most recently. Is this accidental or ?

**Comment by zigurana** [comment]
Hey, I was in the process of cleaning up the branch, an squashing the
commits. I seem to have screwed up the repos (long live Github desktop,
ahum).
I'll need some more time to figure out what needs to happen to clean this
up.

The goal is to get everything related to the UI modes feature in a single
branch,  which is not my main one, so UI_modes_Kiosk_Kid_Full.

That way I am free to work on other features and keep my main branch to
'deliver'  into and run on my bartop.

Let me get back to you once i have everything sorted out.

Btw:if you can recommend any primer on github, Im all ears. Coming from an
ClearCase environment for version SCM, it takes some getting used to.

Gr
Daniel

Op vr 4 mrt. 2016 00:42 schreef Jools Wills <notifications@github.com>:

**Comment by zigurana** [comment]
I'm closing this for now, maybe I'll try to whip it into shape for submission further upstream.

**Pull request closed without merging**